File: plugins/flutter/lib/src/utils/i18n.dart
import 'dart:convert';
import 'package:flutter/services.dart';

class Translations {
  final Map<String, dynamic> _data;
  Translations(this._data);
  Map<String, dynamic> get data => _data;
}

// Default translations (hardcoded English as fallback)
final Map<String, dynamic> _defaultEnglishData = {
  'header': {'subtitle': 'Support'},
  'menu': {
    'title': 'Menu',
    'resetConversation': 'Reset conversation',
    'fullscreen': 'Fullscreen',
    'language': 'Language',
  },
  'buttons': {
    'startConversation': 'Start Conversation',
    'cancel': 'Cancel',
    'reset': 'Reset',
    'save': 'Save',
    'saveChanges': 'Save Changes',
    'add': 'Add',
  },
  'input': {'placeholder': 'Ask a question'},
  'fileUpload': {'fileTypeNotSupported': 'This file type is not supported.'},
  'labels': {'agent': 'Agent', 'you': 'You'},
  'dialog': {
    'resetConversation': {
      'title': 'Reset Conversation',
      'message':
          'This will clear the current conversation history and start a new conversation. Are you sure?',
    },
  },
  'feedback': {'thumbsUp': 'Thumbs up', 'thumbsDown': 'Thumbs down'},
  'thinking': {
    'messages': [
      'Thinking…',
      'Analyzing your question…',
      'Searching knowledge…',
      'Pulling relevant info…',
      'Drafting the answer…',
      'Double-checking details…',
      'Tying it together…',
      'Almost there…',
    ],
  },
  'time': {'justNow': 'Just now', 'today': 'Today', 'yesterday': 'Yesterday'},
  'settings': {'defaultDescription': 'Support'},
  'metadata': {
    'addParameter': 'Add Parameter',
    'editParameter': 'Edit Parameter',
    'defineKeyValue':
        'Define key/value parameters sent as chat metadata.',
  },
};

Translations get defaultTranslations => Translations(_defaultEnglishData);

/// Get translation value by dot-separated key path (e.g. 'header.subtitle')
dynamic getTranslation(String key, Translations translations,
    {String? fallback}) {
  final keys = key.split('.');
  dynamic value = translations.data;
  for (final k in keys) {
    if (value is Map<String, dynamic> && value.containsKey(k)) {
      value = value[k];
    } else {
      return fallback ?? key;
    }
  }
  return value;
}

String getTranslationString(String key, Translations translations,
    {String? fallback}) {
  final value = getTranslation(key, translations);
  if (value is String) return value;
  return fallback ?? key;
}

List<String> getTranslationArray(String key, Translations translations,
    {List<String>? fallback}) {
  final value = getTranslation(key, translations);
  if (value is List) return value.cast<String>();
  return fallback ?? [];
}

/// Resolve language: prop > platform locale > 'en'
String resolveLanguage(String? languageProp) {
  if (languageProp != null && languageProp.isNotEmpty) {
    return languageProp.toLowerCase();
  }
  // Use platform locale - in Flutter we can't easily access this without
  // WidgetsBinding, so default to 'en'
  return 'en';
}

/// Deep merge two maps (source overrides target)
Map<String, dynamic> _deepMerge(
    Map<String, dynamic> target, Map<String, dynamic> source) {
  final output = Map<String, dynamic>.from(target);
  for (final key in source.keys) {
    if (source[key] is Map<String, dynamic> &&
        target[key] is Map<String, dynamic>) {
      output[key] = _deepMerge(
        target[key] as Map<String, dynamic>,
        source[key] as Map<String, dynamic>,
      );
    } else {
      output[key] = source[key];
    }
  }
  return output;
}

Translations mergeTranslations(Map<String, dynamic>? custom,
    {Translations? defaults}) {
  final base = defaults ?? defaultTranslations;
  if (custom == null) return base;
  return Translations(_deepMerge(base.data, custom));
}

/// Load translations for a specific language from bundled JSON asset
Future<Translations> getTranslationsForLanguage(String language) async {
  try {
    final jsonString = await rootBundle
        .loadString('packages/gen_agent_chat/lib/src/l10n/$language.json');
    final data = json.decode(jsonString) as Map<String, dynamic>;
    return Translations(_deepMerge(_defaultEnglishData, data));
  } catch (_) {
    return defaultTranslations;
  }
}

File: plugins/react/src/utils/i18n.ts
// Translation types and utilities
import defaultTranslationsData from './en.json';
import spanishTranslationsData from './es.json';
import frenchTranslationsData from './fr.json';
import germanTranslationsData from './de.json';
import italianTranslationsData from './it.json';
import portugueseTranslationsData from './pt.json';
import chineseTranslationsData from './zh.json';

export interface Translations {
  header: {
    subtitle: string;
  };
  menu: {
    title: string;
    resetConversation: string;
    language: string;
  };
  buttons: {
    startConversation: string;
    cancel: string;
    reset: string;
    save: string;
    saveChanges: string;
    add: string;
  };
  input: {
    placeholder: string;
  };
  labels: {
    agent: string;
    you: string;
  };
  dialog: {
    resetConversation: {
      title: string;
      message: string;
    };
  };
  feedback: {
    thumbsUp: string;
    thumbsDown: string;
  };
  thinking: {
    messages: string[];
  };
  time: {
    justNow: string;
    today: string;
    yesterday: string;
  };
  settings: {
    defaultDescription: string;
  };
  metadata: {
    addParameter: string;
    editParameter: string;
    defineKeyValue: string;
  };
  fileUpload?: {
    fileTypeNotSupported: string;
  };
}

// Default English translations loaded from JSON file
export const defaultTranslations: Translations = defaultTranslationsData as Translations;

/**
 * Get browser language code (e.g., 'en', 'es')
 * Returns the primary language code without region
 */
export function getBrowserLanguage(): string {
  if (typeof window === 'undefined' || !navigator.language) {
    return 'en';
  }
  // Extract primary language code (e.g., 'en' from 'en-US')
  return navigator.language.split('-')[0].toLowerCase();
}

/**
 * Get translation value by key path
 * Supports nested keys like 'header.subtitle' or 'dialog.resetConversation.title'
 */
export function getTranslation(
  key: string,
  translations: Translations,
  fallback?: string
): string | string[] {
  const keys = key.split('.');
  let value: any = translations;

  for (const k of keys) {
    if (value && typeof value === 'object' && k in value) {
      value = value[k as keyof typeof value];
    } else {
      return fallback || key;
    }
  }

  return value;
}

/**
 * Get a string translation (non-array)
 */
export function getTranslationString(
  key: string,
  translations: Translations,
  fallback?: string
): string {
  const value = getTranslation(key, translations, fallback);
  return Array.isArray(value) ? (fallback || key) : value;
}

/**
 * Get an array translation
 */
export function getTranslationArray(
  key: string,
  translations: Translations,
  fallback?: string[]
): string[] {
  const value = getTranslation(key, translations);
  return Array.isArray(value) ? value : (fallback || []);
}

/**
 * Resolve the language to use
 * Priority: prop → browser → 'en'
 */
export function resolveLanguage(languageProp?: string): string {
  if (languageProp) {
    return languageProp.toLowerCase();
  }
  return getBrowserLanguage();
}

/**
 * Merge custom translations with defaults
 * Custom translations override defaults, but don't need to include all keys
 */
export function mergeTranslations(
  custom?: Partial<Translations>,
  defaults: Translations = defaultTranslations
): Translations {
  if (!custom) {
    return defaults;
  }

  // Deep merge function
  const deepMerge = (target: any, source: any): any => {
    const output = { ...target };
    if (isObject(target) && isObject(source)) {
      Object.keys(source).forEach((key) => {
        if (isObject(source[key])) {
          if (!(key in target)) {
            Object.assign(output, { [key]: source[key] });
          } else {
            output[key] = deepMerge(target[key], source[key]);
          }
        } else {
          Object.assign(output, { [key]: source[key] });
        }
      });
    }
    return output;
  };

  return deepMerge(defaults, custom);
}

function isObject(item: any): boolean {
  return item && typeof item === 'object' && !Array.isArray(item);
}

// Spanish translations loaded from JSON file
export const spanishTranslations: Partial<Translations> = spanishTranslationsData as Partial<Translations>;

// French translations loaded from JSON file
export const frenchTranslations: Partial<Translations> = frenchTranslationsData as Partial<Translations>;

// German translations loaded from JSON file
export const germanTranslations: Partial<Translations> = germanTranslationsData as Partial<Translations>;

// Italian translations loaded from JSON file
export const italianTranslations: Partial<Translations> = italianTranslationsData as Partial<Translations>;

// Portuguese translations loaded from JSON file
export const portugueseTranslations: Partial<Translations> = portugueseTranslationsData as Partial<Translations>;

// Chinese translations loaded from JSON file
export const chineseTranslations: Partial<Translations> = chineseTranslationsData as Partial<Translations>;

// Translations map for easy access by language code
export const translationsByLanguage: Record<string, Partial<Translations>> = {
  en: defaultTranslations,
  es: spanishTranslations,
  fr: frenchTranslations,
  de: germanTranslations,
  it: italianTranslations,
  pt: portugueseTranslations,
  zh: chineseTranslations,
};

/**
 * Get translations for a specific language
 */
export function getTranslationsForLanguage(language: string): Translations {
  const langTranslations = translationsByLanguage[language.toLowerCase()];
  if (langTranslations) {
    return mergeTranslations(langTranslations, defaultTranslations);
  }
  return defaultTranslations;
}

